### PR TITLE
chore: Add channel benches

### LINKED ISSUE #6161: Add standalone benches for channels
Benchmark various channel implementations to help us choose which are used when.

Should include benches that use channels in the same ways we use them in vector.

Could be a stand-alone crate with no dependencies on vector.

@MOZGIII worked on some benches here: https://github.com/MOZGIII/rust-benches

### PR DESCRIPTION
**Pull request by MOZGIII**
Closes #6161.

### CODE REVIEW

**Comment by MOZGIII** [comment]
This is a very simple implementation, should work well as a starting point. Ideally, I'd like more people to collaborate on this to add more different benchmarking setups (i.e. multiple threads, buffered/unbuffered channels, etc).

The results I've obtained from these so far are very inconclusive, with the worst case being in the order of ~10 Melems/s and the best being ~15000 Gelem/s.

File: lib/standalone-benches/benches/channels.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            &(batches, batch_size),
+            |b, &(batches, batch_size)| {
+                b.to_async(async_executor::FuturesExecutor).iter_with_setup(
+                    || {
+                        (
+                            tokio02::sync::mpsc::channel(1),
+                            make_data(batches, batch_size),
+                        )
+                    },
+                    |((tx, rx), input)| run_tokio02(tx, rx, input),
```
Comment (jszwedko) [inline lib/standalone-benches/benches/channels.rs:36]: Small nit, I might have just embedded the `run_x` functions in the benchmark. I'm not sure much is gained by splitting out those functions given they are only used in one place.

**Review by jszwedko — COMMENTED** [review]
Nice work! I think having these separate channel benches will be useful.

A few thoughts:

Was there a particular reason you found a need to parameterize the measurement and warm up time? I'm interested by that.

I think it might be good to separate the input parameters here and run them all with a fixed batch size, varying the number of batches, and then a fixed number of batches and varying the batch size as it currently makes it hard to visualize the differences between implementations when two variables changing. You might be able to do that simply by changing the formatting of the benchmark id to be `{}/{}` rather than `{}x{}`. We could also tell criterion to use a logrithmic scale when plotting: https://docs.rs/criterion/0.3.4/criterion/struct.PlotConfiguration.html . I think this would match better with our input scaling.

I also think it'd be useful to test a number of batches of logarithmically starting at 1. Maybe 1, 10, 100, 1000, 10000, 100000.

Our topology seems to use buffered channels:

https://github.com/timberio/vector/blob/3a85bd86fa6a51b635e7282de633ad5699419fff/src/topology/builder.rs#L59

and

https://github.com/timberio/vector/blob/3a85bd86fa6a51b635e7282de633ad5699419fff/src/topology/builder.rs#L118

I'm curious what you think about benching similarly sized buffered channels. We also use fan out.

This is the graph I got when running:

2021-02-02-172542_1727x950_scrot

I'm surprised that the tokio channels seem to perform so much more poorly. I'm wondering if that is expected or if we are "holding them wrong".

Looking at the distribution graphs, it seems like all of the channels have oddly distributed sample distributions when the batch size is 1. In particular tokio 1.1 seems to have a bimodal distribution:

2021-02-02-172752_959x363_scrot

I'm wondering if there is something oddly different between a vector of size 1 and a vector of size 10.

I realize this is contrary to some of what we talked about today regarding warming up of components being benchmarked, but I am wondering what you think of not creating our own batch counts and just letting criterion handle the warm up. Especially with channels that have a buffer size of 1 it's hard to imagine much state being different. I'm curious what you think though.

Just noting also that the use of the newer criterion version seems to have caused some deprecation warnings in the `tracing-limit` benches.

File: lib/standalone-benches/benches/channels.rs (hunk 1 above)
Comment (MOZGIII) [reply lib/standalone-benches/benches/channels.rs:36]: Ah, right. I wanted to see if we macro them out or generalize, but apparently, they are very specific.
I'll do another pass on making things more compact.

**Comment by MOZGIII** [comment]
I'm parametrizing the data size here. My assumption is we might lose a lot of performance in the synchronization (which we do for each sent/received message), and that if we batch our messages together into `Vec`s we'll be getting less overhead (which seems significant), and in the end, we'll ultimately be able to process more data.

So I'm not doing the prewarm there exactly, but measuring the time it takes to pass X amount of batches each containing Y amount of messages. The overall amount of messages passed in then `X * Y`, and that's what's encoded in the `1000x1` part of the run. It makes sense to compare the `AxB` and `CxD` combinations across each other, as the throughput is unified (always elements per second).

I'm still wondering how to make the criterion treat the names properly to compare the values as you illustrated. I tried using `{}/{}` there out of curiosity while I was developing it initially, but it didn't work as in your screenshot :(. I guess we'll have to dig into the criterion source to find out how to make it work.

One thing to note is we should also compare the `Vec<T>` with a single element and a bare `T`, to estimate the core overhead of batching.

File: lib/standalone-benches/benches/channels.rs (hunk 1 above)
Comment (MOZGIII) [reply lib/standalone-benches/benches/channels.rs:36]: Just remembered why I used those - so, the functions are `async`, so each returns a future, and it works with stable rust. The async closures aren't stable yet though, and we'd need those if we want to inline the async calls. Unless I'm not mistaken.

Comment (jszwedko) [reply lib/standalone-benches/benches/channels.rs:36]: I could certainly be missing something, but I think async closures are stable? Using `async move { }` here seems to compile and run fine:

```diff
diff --git a/lib/standalone-benches/benches/channels.rs b/lib/standalone-benches/benches/channels.rs
index 730b0c500..d56b82f67 100644
--- a/lib/standalone-benches/benches/channels.rs
+++ b/lib/standalone-benches/benches/channels.rs
@@ -33,7 +33,13 @@ fn benchmark(c: &mut Criterion) {
                             make_data(batches, batch_size),
                         )
                     },
-                    |((tx, rx), input)| run_tokio02(tx, rx, input),
+                    |((mut tx, mut rx), input)| async move {
+                        for item in input {
+                            tx.send(item).await.unwrap();
+                            let data = rx.recv().await.unwrap();
+                            process_data(data.into_iter());
+                        }
+                    },
                 );
             },
         );
```

**Comment by jszwedko** [comment]
I think this is ready for review now. The purpose of these benches is just to understand the performance characteristics of the different channel implementations and have a playground to try different channel usages. As such, I opted not to wire them into CI. We could even split this into a separate repo if we prefer.

These are the numbers I got (the NxM bit is <number of batches>x<batch size>):

```
group                                          base
-----                                          ----
channels/data_size/async-std 1.9/100000x1      1.00   160.2±70.03ns 581.5 GElem/sec
channels/data_size/async-std 1.9/100000x10     1.00   243.4±44.60ns 3827.1 GElem/sec
channels/data_size/async-std 1.9/100000x100    1.00   278.9±55.85ns 33390.9 GElem/sec
channels/data_size/async-std 1.9/1000x1        1.00     40.9±2.58ns 22.8 GElem/sec
channels/data_size/async-std 1.9/1000x10       1.00     54.9±2.37ns 169.5 GElem/sec
channels/data_size/async-std 1.9/1000x100      1.00    95.1±14.88ns 979.6 GElem/sec
channels/data_size/futures 0.3/100000x1        1.00   151.0±27.37ns 616.6 GElem/sec
channels/data_size/futures 0.3/100000x10       1.00   232.8±37.17ns 4000.0 GElem/sec
channels/data_size/futures 0.3/100000x100      1.00   286.1±71.04ns 32548.9 GElem/sec
channels/data_size/futures 0.3/1000x1          1.00     41.9±2.25ns 22.2 GElem/sec
channels/data_size/futures 0.3/1000x10         1.00     52.5±3.09ns 177.5 GElem/sec
channels/data_size/futures 0.3/1000x100        1.00     98.2±6.55ns 948.5 GElem/sec
channels/data_size/tokio 0.2/100000x1          1.00     19.8±0.16ms  4.8 MElem/sec
channels/data_size/tokio 0.2/100000x10         1.00     21.5±0.23ms 44.4 MElem/sec
channels/data_size/tokio 0.2/100000x100        1.00     29.9±0.47ms 318.9 MElem/sec
channels/data_size/tokio 0.2/1000x1            1.00    184.3±2.35µs  5.2 MElem/sec
channels/data_size/tokio 0.2/1000x10           1.00    192.8±1.72µs 49.5 MElem/sec
channels/data_size/tokio 0.2/1000x100          1.00    206.0±3.41µs 463.0 MElem/sec
channels/data_size/tokio 1.1/100000x1          1.00     23.2±0.30ms  4.1 MElem/sec
channels/data_size/tokio 1.1/100000x10         1.00     24.8±0.23ms 38.4 MElem/sec
channels/data_size/tokio 1.1/100000x100        1.00     35.3±0.41ms 270.5 MElem/sec
channels/data_size/tokio 1.1/1000x1            1.00    224.5±3.10µs  4.2 MElem/sec
channels/data_size/tokio 1.1/1000x10           1.00    227.0±5.05µs 42.0 MElem/sec
channels/data_size/tokio 1.1/1000x100          1.00    240.0±3.78µs 397.3 MElem/sec
```

As Mike hypothesized there are definitely some improvements to be had by batching. We might be able to leverage this knowledge at the topology layer to batch together events between components.

The differences between the number of batches for the same batch size seem odd to me. For example:

```
channels/data_size/async-std 1.9/100000x10     1.00   243.4±44.60ns 3827.1 GElem/sec
channels/data_size/async-std 1.9/1000x10       1.00     54.9±2.37ns 169.5 GElem/sec
```

I'm guessing we get some benefit from those channels continuing to warm as more items are passed through. The tokio channels don't show that benefit.

@blt do you mind taking a look?

**Comment by blt** [comment]
> We could even split this into a separate repo if we prefer.

I think this would be suitable for a separate repo, since we're not interrogating vector's own internals but its dependencies. There's an interesting write-up here, especially if we dig into the channel implementation differences. 

I will review the code properly here in a bit.

**Review by blt — CHANGES_REQUESTED** [review]
I have a few concerns here.

The first concern is two sort of combined: I don't think this benchmark -- which does not test vector internals but vector dependencies -- belongs in the repository and I question, further, the motivation of the work. I am aware that we've suffered throughput regressions when switching channel implementations in the past. Is the motivation here to find the fastest channel for vector? If that's the case, I think all-up effects will dominate over the kinds of things we can measure in a microbench and, to my knowledge, our work in vector-test-harness hasn't pointed to our channels being a bottleneck in a way we can run experiments with yet. That said, out of this repo a fair comparison between the various channels _and_ a deep dive into their implementations to explain observed differences would be valuable to the community. That I do support, which leads to my second concern. 

As noted in the intro to this PR the work is preliminary, lacking tests for multiple thread counts etc. While we can extend these tests to cover such cases I don't think criterion is the right tool for this job. Once we have the up/down from criterion what do we do next? If we want to understand why the channel internals lead to our observed results we'll need to write a binary target that runs the channel in a similar rig, feed that through `perf` and generate a flamegraph or diagnostic tool. That'll be especially true if we need to understand the implication of atomics, something criterion hints at but can't point to. It strikes me that what we're after here is more about _integration_ behavior -- potentially in a way that is indicative of our use in vector, but maybe not -- and less about the behavior of this or that function. With criterion I'd want to answer questions like "What's the profile of a `send` if its channel has space versus when the channel has to resize or when it is full?" and not "How quickly can I send 100Gb worth of some struct through a buffered channel if I dedicate one thread for the sender, one for the receiver? What about two for the receiver? Four? Eight?"

**Comment by jszwedko** [comment]
Thanks for the thoughts @blt! I'm realizing I didn't provide enough context for this PR. We talked a bit offline, but expanding here:

These benches are really just supposed to be for research and experimentation. I agree with you that we shouldn't make changes there until we have more of the improved test harness in-place and can point to channels as being a significant contributor to overall performance.

Some inline responses:

> The first concern is two sort of combined: I don't think this benchmark -- which does not test vector internals but vector dependencies -- belongs in the repository 

Agreed, I'll pull this out into a separate repo.

> Is the motivation here to find the fastest channel for vector?

I think the motivation is just to understand the channel profiles. It could help us choose which channels are the most appropriate for vector use-cases though in the various places we use them. It seems possible that some places will even prefer different channels than others.

> As noted in the intro to this PR the work is preliminary, lacking tests for multiple thread counts etc. While we can extend these tests to cover such cases I don't think criterion is the right tool for this job. Once we have the up/down from criterion what do we do next? If we want to understand why the channel internals lead to our observed results we'll need to write a binary target that runs the channel in a similar rig, feed that through perf and generate a flamegraph or diagnostic tool. 

My limited understanding is that different channels have made different implementation choices optimizing for different things. I think these benches could just give us some more info to help decide which channels to use where depending on the use-case. I don't expect to optimize the channel implementations themselves. I can see that it would be helpful to have a deeper understanding of their implementation though to understand _why_ they behave the way they do if we find them to be a bottleneck.

> It strikes me that what we're after here is more about integration behavior -- potentially in a way that is indicative of our use in vector, but maybe not -- and less about the behavior of this or that function. With criterion I'd want to answer questions like "What's the profile of a send if its channel has space versus when the channel has to resize or when it is full?" and not "How quickly can I send 100Gb worth of some struct through a buffered channel if I dedicate one thread for the sender, one for the receiver? What about two for the receiver? Four? Eight?"

I agree that this work is rudimentary and doesn't explore these other characteristics that would be helpful to know like behavior with multiple readers/writers.

**Comment by jszwedko** [comment]
@blt I pulled these out into a separate repo. Do you mind taking a look? https://github.com/timberio/rust-benches/pull/1

**Pull request closed without merging**